src/mpid/ch4/netmod/ucx/ucx_impl.h

```diff
@@ -101,8 +101,8 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_UCX_get_source(uint64_t match_bits)
 MPL_STATIC_INLINE_PREFIX bool MPIDI_UCX_is_reachable_target(int rank, MPIR_Win * win,
                                                             MPIDI_winattr_t winattr)
 {
-    /* zero win target does not have rkey. */
-    return (winattr & MPIDI_WINATTR_NM_REACHABLE) && MPIDI_UCX_WIN_INFO(win, rank).rkey != NULL;
+    /* unmapped win target does not have rkey. */
+    return (winattr & MPIDI_WINATTR_NM_REACHABLE) || MPIDI_UCX_WIN_INFO(win, rank).rkey != NULL;
 }
 
```

Comment:
Liked this commit. Can we do the same optimization for OFI/RMA? In OFI, we currently force AM for all processes if any of them fails to register memory region.

Reply:
So we need some way to indicate that the `winfo` object for a particular rank is invalid. It looks like we have `disp_unit`, `mr_key`, and `base_ptr`. Is there any special value we can set to indicate that a target is unreachable? I don't see anything in the libfabric man pages at a glance. Perhaps we add a boolean into the struct? Locally I know that my MR failed, so I set `MPIDI_OFI_WIN(win).mr = NULL;`

Reply:
Hmm, actually this is not as simple as the UCX code. I'm having trouble following the logic between the different cases such as:
```
MPL_STATIC_INLINE_PREFIX uintptr_t MPIDI_OFI_winfo_base(MPIR_Win * w, int rank)
{
    if (!MPIDI_OFI_ENABLE_MR_VIRT_ADDRESS)
        return 0;
    else
        return MPIDI_OFI_WIN(w).winfo[rank].base;
}

MPL_STATIC_INLINE_PREFIX uint64_t MPIDI_OFI_winfo_mr_key(MPIR_Win * w, int rank)
{
    if (!MPIDI_OFI_ENABLE_MR_PROV_KEY)
	return MPIDI_OFI_WIN(w).mr_key;
    else
        return MPIDI_OFI_WIN(w).winfo[rank].mr_key;
}
```
Not sure how to adjust these to deal with unsuccessful mappings.

Reply:
I see. OFI does some MR-related optimizations which make the logic overcomplex. Let's just do not change the OFI part in this PR.

Reply:
OK, sounds good. Let me fix the current failures and this should be ready for review.